model/pipeline/src/main/proto/beam_runner_api.proto

```diff
@@ -855,10 +855,21 @@ message StandardCoders {
     //     BOOLEAN:   beam:coder:bool:v1
     //     BYTES:     beam:coder:bytes:v1
     //   ArrayType:   beam:coder:iterable:v1 (always has a known length)
-    //   MapType:     not yet a standard coder (BEAM-7996)
+    //   MapType:     not a standard coder, specification defined below.
     //   RowType:     beam:coder:row:v1
     //   LogicalType: Uses the coder for its representation.
     //
+    // The MapType is encoded by:
+    //   - An INT32 representing the size of the map (N)
+    //   - Followed by N interleaved keys and values, encoded with their
+    //     corresponding coder.
+    //
+    // Nullable types in container types (ArrayType, MapType) are encoded by:
+    //   - A one byte null indicator, 0x00 for null values, or 0x01 for present
+    //     values.
+    //   - For present values the null indicator is followed by the value
+    //     encoded with it's corresponding coder.
+    //
```

Comment:
FYI @lostluck - previously the encoding for MapType and for nullable types within containers was not documented.

Reply:
Ack! Thanks!

Reply:
Is this just the encoding of a nullable type? (Why does it have to be called out specially?)

Reply:
For maps specifically, do we want to allow null keys? Is it valuable to have null values (as distinct from just not present)? I might lean towards disallowing nulls and then possibly allowing it in the future if we have good reason to, which will be forward compatible.

Reply:
This is specifically for nullable types that are elements of an Array or keys/values of a Map. For rows we nulls in a separate bitmask:
https://github.com/apache/beam/blob/587dde57cbb2b0095a1fa04b59798d1b62c66f18/model/pipeline/src/main/proto/beam_runner_api.proto#L837-L843

Java schemas will already let you use nullable types for keys and values in Maps. I doubt anyone is relying on it.. but there's a risk if we disallow it now.

Reply:
Just so I understand it right, at the schema level, is nullability field is a property of the schema (rather than having an optional type)? So there's no way to declare a map as not (potentially) having null keys/values?

Reply:
That's right. Nullability is represented on the FieldType: https://github.com/apache/beam/blob/d5a6c2f0ceb7b3e609ba8d97e523b3bded7d0459/model/pipeline/src/main/proto/schema.proto#L47

We _could_ document that schemas should not accept null keys/values in `MapType` and reject them in the SDKs. But there would still be an unused nullable field on the key and value FieldType:
https://github.com/apache/beam/blob/d5a6c2f0ceb7b3e609ba8d97e523b3bded7d0459/model/pipeline/src/main/proto/schema.proto#L79-L82

Reply:
So the verdict here is to ignore the nullability field of a FieldType when it's nested in a Map or Array?

That's... unfortunate. Must we bend over backwards to maintain compatibility with Java's previous encoding, which wasn't officially a schema encoding until this PR? It seems very strange that we're already relying on unspecified behaviour.

Reply:
Also, this requirement means that *all* values must be pointers (or reference types) in Go as ordinary primitives cannot be nullable. That feels very strange.

Reply:
@robertwb Your comment fed my own misunderstanding. It is possible to declare a map as not having null/keys values.
Technically, there's no reason that the SDK can't use a non-nullable containing version of the container if the Key and Value components are not themselves marked as nillable. IIRC, the Java SDK *could* converted ImmutableMaps or similar into just non-nullable Key and non-nullable Value types. 
The issue as I'm understanding it is that the limitation is on the SDK Language side, rather than the schema specification side, as discussed the schemas fields can individually have their nullable bits set.
Eg. Go doesn't have this ambiguity for map types.
On the other hand, in Go, Iterable/array types which will be represented by slices *will* have this ambiguity when used as a field, as they can be nil, and could also still be pointers to said reference types. That ambiguity is well known enough that pointers to reference types (maps, slices, chans..) are strongly discouraged in idiomatic Go.

Reply:
@lostluck The issue here is what kinds of objects the SDK must accept when receiving these from a foreign SDK. If nulls are allowed, you cannot let the element type be (say) a `map[str, int]` in go. 

It looks like there is a way to specify in the schema whether the keys/values could be null (or explicitly disallow it for either or both). The encoding, on the other hand, always has this leading bit set (regardless of whether the schema allows null values), but is harder to change for backwards compatibility reasons. If this is the case, I'm OK with that.

Reply:
Nevermind, we set this prefix depending on the bit in the schema. All is good.